### DISCUSSION TITLE
Stop storing the full IP address and user agent

### DISCUSSION THREAD

**Discussion by stevygee** [walinejs/waline#1485] [Q&A] [answered]
Is it necessary to store the full IP address and user agent in the database?

I would prefer to store as little personal information as possible. In Europe under the GDPR law, the IP address is considered to be personal information as well.

It would be great to have a setting for 1) pseudonymization of IP addresses and 2) for turning off storage of the user agent. Of course, for information already stored in the database, it would have to be removed as well, either manually or automatically, once the setting(s) are activated.

**Comment by lizheming** [comment, marked as answer]
You can use `preSave` hooks to do it if you want to stop store IP and UA address when post comment. https://waline.js.org/en/reference/server.html#presave-comment

```js
// index.js 
const Waline = require('@waline/vercel');
module.exports = Waline({
  preSave(data) {
    delete data.ua;
    // store ip 192.168.1.1 with mask like 192.*.*.1
    data.ip = data.ip.replace(/\.\d+\.\d+\./, '.*.*.');
    // or you can use delete data.ip to stop store it. 
    // delete data.ip;
  }
});
```

If you want to clean stored ip and ua, you can go to leancloud dashboard to clear it like following:

![image](https://user-images.githubusercontent.com/424491/194526472-90fe8462-85fd-4203-9762-f165ed81a478.png)

**Reply by Mister-Hope** [reply]
We can add a special page in docs to mention privacy.

**Reply by stevygee** [reply]
That works great! Thank you!

**Comment by Suleman-Elahi** [comment]
Is this the reason why comment box doesn't show in EU countries ?

**Reply by lizheming** [reply]
nope, could you show me url which can reproduce the problem. I'll check what's wrong.

**Reply by Suleman-Elahi** [reply]
https://www.ilovefreesoftware.com/11/webware/convert-screen-recording-to-gif-for-free-online-tool.html
https://www.fosstools.net/how-to-run-docker-commands-without-sudo-on-ubuntu-debian-arch-linux

In EU the comment box doesn't show up. I tested virtually from Germany. Also, most of the comments I receive from Turkey end up in WordPress's default comment system while other comments show in Waline Admin Panel.

**Reply by Suleman-Elahi** [reply]
Turns out it's the problem with Facebook comments as well: https://developers.facebook.com/docs/plugins/comments/

**Reply by lizheming** [reply]
We have no action for EU region policy. If the comment box doesn't show up, the biggest possible is the local network. Sorry I can't test EU by myself right now. Cloud you show me Chrome Devtool network tab screenshot after visit https://www.fosstools.net/how-to-run-docker-commands-without-sudo-on-ubuntu-debian-arch-linux like this:

![image](https://github.com/walinejs/waline/assets/424491/9d8ad80f-f6a1-4cc5-b373-a083c4336f68)

**Reply by Suleman-Elahi** [reply]
Sorry for the trouble but now it seems to be working. I checked earlier and it was not working. But after trying from another server in Germany, it works.

**Reply by lizheming** [reply]
OK, that's so great there are no problems. If there are any problems after, you can create a new discussion report it and we'll check it asap.
